### PR TITLE
Apply dropout to decoder

### PR DESCRIPTION
**Pull request by JohnGiorgi**
Adds dropout to the target embeddings and the inputs of the decoder. I found that this helped improve the models performance slightly. Additionally, I made the default weight decay for the decoder side params `0.01`, and added a non-linearity to the projected decoder inputs.

I have confirmed that this leads to small boosts in performance on both BC5CDR and GDA. Interestingly, I found that I had to lower the batch size when using entity hinting (from 16 --> 4). I don't have a good hypothesis for why this is (maybe examples with entity hinting are less noisy and therefore can be trained with a smaller batch size?).

### CODE REVIEW

File: seq2rel/models/copynet_seq2rel.py [hunk 1]
```diff
@@ -194,6 +201,42 @@ def forward(
...
+        # shape: (group_size, source_sequence_length, encoder_output_dim)
+        encoder_outputs_mask = state["source_mask"]
+        # shape: (group_size, target_embedding_dim)
+        embedded_input = self._target_embedder(last_predictions)
+        embedded_input = self._dropout(embedded_input)
+        # shape: (group_size, source_sequence_length)
+        attentive_weights = self._attention(
+            state["decoder_hidden"], state["encoder_outputs"], encoder_outputs_mask
+        )
+        # attentive_weights = self._dropout(attentive_weights)
```
Comment (JohnGiorgi) [inline seq2rel/models/copynet_seq2rel.py]:
```suggestion
```
